### PR TITLE
[CARBONDATA-1954] [Pre-Aggregate] CarbonHiveMetastore updated while dropping the Pre-Aggregate table & code refactored

### PR DESCRIPTION
**Pull request by rahulforallp**
**This PR contains :**
**1.** To update CarbonHiveMetastore similar function was already there . Removed duplicate function defination and updated the caller.
**2.** code refactored so that during droping a  pre-aggregate table only metadata will be deleted if processMetadata() is called.

 - [x] Any interfaces changed? No
 
 - [X] Any backward compatibility impacted? No
 
 - [X] Document update required? No

 - [X] Testing done - Yes, test case added
     
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2633/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2447/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1227/

**Comment by rahulforallp** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2645/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala [hunk 1]
```diff
@@ -76,6 +78,24 @@ class TestDataMapCommand extends QueryTest with BeforeAndAfterAll {
     assert(dataMapSchemaList.get(2).getChildSchema.getTableName.equals("datamaptest_datamap3"))
   }
 
+  test("check hivemetastore after drop datamap") {
+    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.ENABLE_HIVE_SCHEMA_META_STORE,
```
Comment (gvramana) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala]: reset this property

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala [hunk 2]
```diff
@@ -76,6 +78,24 @@ class TestDataMapCommand extends QueryTest with BeforeAndAfterAll {
...
+    sql("drop table if exists hiveMetaStoreTable")
+    sql("create table hiveMetaStoreTable (a string, b string, c string) stored by 'carbondata'")
+    sql(
+      "create datamap datamap_hiveMetaStoreTable on table hiveMetaStoreTable using 'preaggregate' dmproperties('key'='value') as select count(a) from hiveMetaStoreTable")
+
+    val table = CarbonMetadata.getInstance().getCarbonTable("default", "hiveMetaStoreTable")
+    var dataMapSchemaList = table.getTableInfo.getDataMapSchemaList
+    assert(dataMapSchemaList.size() == 1)
+
+    sql("drop datamap if exists datamap_hiveMetaStoreTable on table hiveMetaStoreTable")
```
Comment (gvramana) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala]: drop main table also

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonHiveMetaStore.scala [hunk 3]
```diff
@@ -194,6 +194,11 @@ class CarbonHiveMetaStore extends CarbonFileMetastore {
       newTablePath)
     val dbName = oldTableIdentifier.getDatabaseName
     val tableName = oldTableIdentifier.getTableName
+    val schemaParts = CarbonUtil.convertToMultiGsonStrings(wrapperTableInfo, "=", "'", "")
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonHiveMetaStore.scala]: Not required this function use above function , updateHiveMetaStoreForAlter

**Comment by gvramana** [comment]
change PR title and issue title to describe the problem

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2513/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1296/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2681/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/DataMapListeners.scala [hunk 4]
```diff
@@ -65,7 +65,7 @@ object DropDataMapPostListener extends OperationEventListener {
           Some(dataMapSchema.get.getRelationIdentifier.getDatabaseName),
           dataMapSchema.get.getRelationIdentifier.getTableName,
           dropChildTable = true
-        ).run(sparkSession)
+        ).processMetadata(sparkSession)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/DataMapListeners.scala]: I think it is better to move this call to CarbonDropDataMapCommand.processMetadata

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala [hunk 5]
```diff
@@ -136,6 +137,12 @@ case class CarbonDropDataMapCommand(
     // delete the table folder
     val tableIdentifier = CarbonEnv.getIdentifier(databaseNameOp, tableName)(sparkSession)
     DataMapStoreManager.getInstance().clearDataMap(tableIdentifier, dataMapName)
+    CarbonDropTableCommand(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala]: Beside CarbonDropDataMapCommand.scala, CarbonDropTableCommand.scala also has problem that processMetadata is incorrect if there are datamaps associated with the fact table. Please modify that in this PR also.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala [hunk 6]
```diff
@@ -76,6 +78,31 @@ class TestDataMapCommand extends QueryTest with BeforeAndAfterAll {
...
+
+    }
+    finally {
+      sql("drop table hiveMetaStoreTable")
+      CarbonProperties.getInstance()
+        .addProperty(CarbonCommonConstants.ENABLE_HIVE_SCHEMA_META_STORE,
+          CarbonCommonConstants.ENABLE_HIVE_SCHEMA_META_STORE_DEFAULT)
+    }
+  }
+
```
Comment (jackylk) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala]: Beside CarbonDropDataMapCommand.scala, CarbonDropTableCommand.scala also has problem that processMetadata is incorrect if there are datamaps associated with the fact table. Please modify that in this PR also and add testcase

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/DataMapListeners.scala (hunk 4 above)
Comment (rahulforallp) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/DataMapListeners.scala]: @jackylk  code is moved to CarbonDropDataMapCommand

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2548/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1323/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2718/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2743/

**Comment by rahulforallp** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2745/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2580/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2588/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2751/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1353/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2780/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2618/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1383/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1384/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2619/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2782/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2620/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1385/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala (hunk 6 above)
Comment (rahulforallp) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala]: CarbonDropDataMapCommand also  modified and test case added.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1404/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2798/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2639/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2800/

**Comment by ravipesala** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1410/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2646/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala [hunk 7]
```diff
@@ -76,6 +78,58 @@ class TestDataMapCommand extends QueryTest with BeforeAndAfterAll {
     assert(dataMapSchemaList.get(2).getChildSchema.getTableName.equals("datamaptest_datamap3"))
   }
 
+  test("check hivemetastore after drop datamap") {
+    try {
+      CarbonProperties.getInstance()
+        .addProperty(CarbonCommonConstants.ENABLE_HIVE_SCHEMA_META_STORE,
+          "true")
+      sql("drop datamap if exists datamap_hiveMetaStoreTable on table hiveMetaStoreTable")
```
Comment (xubo245) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datamap/TestDataMapCommand.scala]: where is hiveMetaStoreTable?

**Pull request closed without merging**